File: apps/mobile/native/android/com/j11/looper/mobile/repos/ApiUtils.kt
package com.j11.looper.mobile.repos

import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URI
import java.net.URL

enum class ConvexFunctionKind(val path: String) {
    QUERY("query"),
    MUTATION("mutation"),
    ACTION("action"),
}

data class HttpResponse(
    val status: Int,
    val body: String,
)

fun postJsonSync(
    urlString: String,
    payload: JSONObject,
    authorization: String? = null,
    extraHeaders: Map<String, String> = emptyMap(),
): HttpResponse? = BlockingPost.execute(
    PostRequest(
        destination = urlString,
        content = payload.toString().toByteArray(),
        headers = postHeaders("application/json", authorization, extraHeaders),
    ),
)

fun postBytesSync(
    urlString: String,
    data: ByteArray,
    contentType: String,
    headers: Map<String, String> = emptyMap(),
): HttpResponse? = BlockingPost.execute(
    PostRequest(
        destination = urlString,
        content = data,
        headers = postHeaders(contentType, null, headers),
    ),
)

fun invokeHandlerSync(
    config: RepoConfig,
    kind: ConvexFunctionKind,
    name: String,
    args: JSONObject,
): JSONObject? = invokeConvexSync(config, kind, name, args) as? JSONObject

fun invokeHandlerRawValueSync(
    config: RepoConfig,
    kind: ConvexFunctionKind,
    name: String,
    args: JSONObject,
): Any? = invokeConvexSync(config, kind, name, args)

fun uploadFileSync(
    urlString: String,
    data: ByteArray,
    contentType: String,
    configuredConvexUrl: String? = null,
): String? {
    val destination = UploadDestination.resolve(urlString, configuredConvexUrl)
    val response = postBytesSync(destination, data, contentType) ?: return null
    if (!response.successful) {
        warn("upload failed: HTTP ${response.status} ${response.body.take(LOG_BODY_LIMIT)}")
        return null
    }
    return try {
        val identifier = JSONObject(response.body).opt("storageId") as? String
        if (identifier == null) warn("upload failed: no storageId in response")
        identifier
    } catch (error: Exception) {
        warn("upload failed: ${error.message}")
        null
    }
}

private fun invokeConvexSync(
    config: RepoConfig,
    kind: ConvexFunctionKind,
    name: String,
    args: JSONObject,
): Any? = try {
    val call = ConvexWire.call(config, kind, name, args)
    val response = postJsonSync(call.destination, call.payload, call.authorization) ?: return null
    ConvexWire.readValue(name, response)
} catch (error: Exception) {
    warn("$name failed: ${error.message}")
    null
}

internal data class ConvexCall(
    val destination: String,
    val payload: JSONObject,
    val authorization: String,
)

internal object ConvexWire {
    fun call(
        config: RepoConfig,
        kind: ConvexFunctionKind,
        name: String,
        args: JSONObject,
    ): ConvexCall {
        val envelope = JSONObject()
            .put("path", name)
            .put("format", "convex_encoded_json")
            .put("args", JSONArray().put(args))
        return ConvexCall(
            destination = "${config.functionUrl}/api/${kind.path}",
            payload = envelope,
            authorization = "Bearer ${config.idToken}",
        )
    }

    fun readValue(operation: String, response: HttpResponse): Any? {
        if (!response.successful) {
            warn("$operation failed: HTTP ${response.status} ${response.body.take(LOG_BODY_LIMIT)}")
            return null
        }
        val envelope = JSONObject(response.body)
        if (envelope.optString("status") != "success") {
            val fallback = "$operation failed"
            warn("$operation failed: ${envelope.optString("errorMessage", fallback)}")
            return null
        }
        return envelope.opt("value")
    }
}

internal object UploadDestination {
    fun resolve(uploadUrl: String, configuredConvexUrl: String?): String {
        if (configuredConvexUrl.isNullOrBlank()) return uploadUrl
        return runCatching {
            val upload = URI(uploadUrl)
            val configured = URI(configuredConvexUrl)
            if (!upload.host.isLoopback() || configured.host.isLoopback()) return uploadUrl
            URI(
                configured.scheme,
                upload.userInfo,
                configured.host,
                configured.port,
                upload.path,
                upload.query,
                upload.fragment,
            ).toString()
        }.getOrDefault(uploadUrl)
    }
}

private data class PostRequest(
    val destination: String,
    val content: ByteArray,
    val headers: Map<String, String>,
)

private object BlockingPost {
    fun execute(request: PostRequest): HttpResponse? = try {
        val connection = URL(request.destination).openConnection() as HttpURLConnection
        connection.requestMethod = "POST"
        request.headers.forEach(connection::setRequestProperty)
        connection.doOutput = true
        connection.outputStream.use { stream -> stream.write(request.content) }
        val status = connection.responseCode
        val responseStream = if (status in SUCCESS_RANGE) {
            connection.inputStream
        } else {
            connection.errorStream
        }
        val body = responseStream?.bufferedReader()?.use { it.readText() }.orEmpty()
        connection.disconnect()
        HttpResponse(status, body)
    } catch (error: Exception) {
        warn("HTTP call failed: ${error.message}")
        null
    }
}

private fun postHeaders(
    contentType: String,
    authorization: String?,
    overrides: Map<String, String>,
): Map<String, String> {
    val result = LinkedHashMap<String, String>(overrides.size + 2)
    result["Content-Type"] = contentType
    authorization?.takeIf(String::isNotBlank)?.let { result["Authorization"] = it }
    result.putAll(overrides)
    return result
}

private val HttpResponse.successful: Boolean
    get() = status in SUCCESS_RANGE

private fun String?.isLoopback(): Boolean = when (this?.lowercase()) {
    "localhost", "127.0.0.1", "::1" -> true
    else -> false
}

private fun warn(message: String) {
    Log.w(LOG_TAG, message)
}

private val SUCCESS_RANGE = 200..299
private const val LOG_BODY_LIMIT = 200
private const val LOG_TAG = "LooperRepo"

File: apps/mobile/targets/keyboard/Utils/ApiUtils.swift
import Foundation

enum ConvexFunctionKind: String {
  case query
  case mutation
  case action
}

enum ApiError: Error, LocalizedError {
  case invalidURL
  case httpError(Int, String)
  case convexError(String)
  case parseError

  var errorDescription: String? {
    switch self {
    case .invalidURL:
      return "Invalid function URL"
    case .httpError(let status, let responseBody):
      return "HTTP \(status): \(responseBody.prefix(200))"
    case .convexError(let explanation):
      return explanation
    case .parseError:
      return "Could not parse response"
    }
  }
}

typealias ApiTransport = (URLRequest) async throws -> (Data, URLResponse)

private struct ConvexInvocation {
  let config: RepoConfig
  let kind: ConvexFunctionKind
  let path: String
  let arguments: [String: Any]

  func request() throws -> URLRequest {
    guard let endpoint = URL(string: config.functionUrl + "/api/" + kind.rawValue) else {
      throw ApiError.invalidURL
    }
    var request = URLRequest(url: endpoint)
    request.httpMethod = "POST"
    request.httpBody = try JSONSerialization.data(withJSONObject: [
      "path": path,
      "format": "convex_encoded_json",
      "args": [arguments],
    ])
    request.setValue("Bearer " + config.idToken, forHTTPHeaderField: "Authorization")
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    return request
  }
}

private enum ApiResponse {
  static func successfulData(_ data: Data, _ response: URLResponse) throws -> Data {
    let status = (response as? HTTPURLResponse)?.statusCode ?? -1
    guard (200...299).contains(status) else {
      throw ApiError.httpError(status, String(decoding: data, as: UTF8.self))
    }
    return data
  }

  static func convexValue(from data: Data, functionName: String) throws -> Any? {
    guard let envelope = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
      throw ApiError.parseError
    }
    guard envelope["status"] as? String == "success" else {
      let fallback = functionName + " failed"
      throw ApiError.convexError(envelope["errorMessage"] as? String ?? fallback)
    }
    return envelope["value"]
  }

  static func storageIdentifier(from data: Data) throws -> String {
    let decoded = try JSONSerialization.jsonObject(with: data)
    guard
      let fields = decoded as? [String: Any],
      let storageIdentifier = fields["storageId"] as? String
    else {
      throw ApiError.parseError
    }
    return storageIdentifier
  }
}

private let liveApiTransport: ApiTransport = { request in
  try await URLSession.shared.data(for: request)
}

func invokeConvexValue(
  config: RepoConfig,
  kind: ConvexFunctionKind,
  name: String,
  args: [String: Any],
  transport: @escaping ApiTransport
) async throws -> Any? {
  let operation = ConvexInvocation(config: config, kind: kind, path: name, arguments: args)
  let (data, response) = try await transport(operation.request())
  return try ApiResponse.convexValue(
    from: ApiResponse.successfulData(data, response),
    functionName: name
  )
}

func invokeHandler(
  config: RepoConfig,
  kind: ConvexFunctionKind,
  name: String,
  args: [String: Any]
) async throws -> [String: Any] {
  try await invokeHandler(
    config: config,
    kind: kind,
    name: name,
    args: args,
    transport: liveApiTransport
  )
}

func invokeHandler(
  config: RepoConfig,
  kind: ConvexFunctionKind,
  name: String,
  args: [String: Any],
  transport: @escaping ApiTransport
) async throws -> [String: Any] {
  let value = try await invokeConvexValue(
    config: config,
    kind: kind,
    name: name,
    args: args,
    transport: transport
  )
  guard let object = value as? [String: Any] else { throw ApiError.parseError }
  return object
}

func invokeHandlerRawValue(
  config: RepoConfig,
  kind: ConvexFunctionKind,
  name: String,
  args: [String: Any]
) async throws -> Any {
  guard
    let value = try await invokeConvexValue(
      config: config,
      kind: kind,
      name: name,
      args: args,
      transport: liveApiTransport
    )
  else {
    throw ApiError.parseError
  }
  return value
}

func uploadFile(to urlString: String, data: Data, contentType: String) async throws -> String {
  try await uploadFile(
    to: urlString,
    data: data,
    contentType: contentType,
    transport: liveApiTransport
  )
}

func uploadFile(
  to urlString: String,
  data: Data,
  contentType: String,
  transport: @escaping ApiTransport
) async throws -> String {
  guard let endpoint = URL(string: urlString) else { throw ApiError.invalidURL }
  var request = URLRequest(url: endpoint)
  request.httpMethod = "POST"
  request.httpBody = data
  request.setValue(contentType, forHTTPHeaderField: "Content-Type")

  let (responseData, response) = try await transport(request)
  let successfulData = try ApiResponse.successfulData(responseData, response)
  return try ApiResponse.storageIdentifier(from: successfulData)
}

func invokeHandlerFireAndForget(
  config: RepoConfig,
  kind: ConvexFunctionKind,
  name: String,
  args: [String: Any]
) {
  Task {
    do {
      _ = try await invokeHandler(config: config, kind: kind, name: name, args: args)
    } catch {
      NSLog("[LooperKB] %@ failed: %@", name, error.localizedDescription)
    }
  }
}
